### ISSUE TITLE
a thread-safe problem

### ISSUE THREAD

**Issue by seedeed** [kilim/kilim#67]
Hi, I read the code of writeHeader method of kilim.http.HttpResponse class recently. And I found that a static field gmtdf with type java.text.SimpleDateFormat is used for writing HTTP Date header. But as is known to me, the format method of java.text.SimpleDateFormat is not safe in concurrent environment without any synchronization. In order to solve this problem, I think creating a SimpleDateFormat instance for every HttpResponse.writeHeader call, adding an external synchronized block for SimpleDateFormat.format call, or using ThreadLocal for field gmtdf will all be helpful. Anyway, I can submit a PR for this problem after a confirm.

**Comment by nqzero** [comment]
Yeah, that looks like a problem. We're also calling `new Date()` on every response which is slow (my understanding is that for performance most servers cache this header, eg see [the first bullet point in this blog post](https://www.techempower.com/blog/2019/07/09/framework-benchmarks-round-18/)).

In general thread locals are risky to use with woven code but that method isn't Pausable so it should work fine here.

The "Kilim" way to solve this is to have an actor that does the formatting and caches the result, with a timer to update it once per second.

Long term, we should probably benchmark the two approaches and see how they impact performance. But for now, let's just fix the problem. If you're willing to tackle either one, that would be great.

I'm not sure how familiar you are with using Kilim, but if you're new to it, the actor approach would be a great chance to learn and I'll be available to help

And again, thanks for the bug report, it's greatly appreciated !

**Comment by seedeed** [comment]
Yeah, it seems that the 'Kilim' way will be more high-performance. I will try it.

**Comment by sriram-srinivasan** [comment]
Have you benchmarked java.time.format? It is supposed to be thread-safe and much much faster than the original SimpleDateFormatter.    If I am not mistaken  the Java folks incorporated ideas from the Joda time project.

**Comment by seedeed** [comment]
Sorry for my late reply. After my simple benchmark about SimpleDateFormat and DateTimeFormatter from jdk8. It seems that DateTimeFormatter from jdk8 has a better performance and is documented as safe-thread. 

benchmark result: 
```
Benchmark                                           Mode  Cnt        Score        Error  Units
DateFormatBenchmark.testDateTimeFormatterFromJdk8  thrpt   10  3050330.237 ±  40854.756  ops/s
DateFormatBenchmark.testSimpleDateFormat           thrpt   10  2478605.973 ± 402731.544  ops/s
```

benchmark code:
```java
package cn.seedeed.jmh;

import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.concurrent.TimeUnit;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Measurement;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.Setup;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.annotations.Threads;
import org.openjdk.jmh.annotations.Warmup;


@BenchmarkMode(Mode.Throughput)
@Warmup(iterations = 2, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 10, time = 3, timeUnit = TimeUnit.SECONDS)
@Threads(1)
@Fork(1)
@OutputTimeUnit(TimeUnit.SECONDS)
@State(Scope.Thread)
public class DateFormatBenchmark {
	
	private SimpleDateFormat sdf;
	private Date date;
	private DateTimeFormatter dtFormatter;
	private LocalDateTime localDt;
	
	@Setup
	public void setup() {
		sdf = new SimpleDateFormat("EEE, dd MMM yyyy hh:mm:ss");
		date = new Date();
		dtFormatter = DateTimeFormatter.ofPattern("EEE, dd MMM yyyy hh:mm:ss");
		localDt = LocalDateTime.now();
	}

	@Benchmark
	public void testSimpleDateFormat() {
		sdf.format(date);
	}
	
	@Benchmark
	public void testDateTimeFormatterFromJdk8() {
		localDt.format(dtFormatter);
	}
}

```
